### PR TITLE
#451 fix scala-js deps

### PR DESCRIPTION
**Pull request by massimosiani**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -90,9 +90,8 @@ val baseSettings = Seq(
   version := buildVersion,
   ThisBuild / scalaVersion := scala213,
   crossScalaVersions := crossVersionAll,
-  crossVersion := CrossVersion.binary,
```
Comment (massimosiani) [inline build.sbt:93]: That's why the cross dependency.
I published the artifacts locally and their dependencies looks fine. I'll work more on the js tests, but the publishing of the jvm part should be fine again.

**Comment by erwan** [comment]
@massimosiani so do you think it's ready to merge? I'll trust you on this one because I don't know much about ScalaJS!

**Comment by MirkoManojlovic** [comment]
Hello everyone, first of all I want to thank you for the awesome project, this is really helping a lot :) 

I'm using it on the backend, but as other have mentions I'm building a frontend with ScalaJ and want to include it there as well.
@massimosiani I've checked out your `build-js` branch and when I'm on [fix scala-js deps](https://github.com/jwt-scala/jwt-scala/pull/460/commits/394f1432fc50afce189b1e94d7fb9dcfb280409b) commit I've tried publishing everything locally and it works fine, I've included the lib on my ScalaJs frontend and everything works, so I want to ask what was the reason for this [remove JS from the build](https://github.com/jwt-scala/jwt-scala/pull/460/commits/e89929c5a2c2cc9c4884573468e9374458183462) commit?

I'm willing to help out in anyway, I have some spare time on my hands so if I can do something please let me know :) 

Once again thanks for the awesome project :)

**Comment by sbrunk** [comment]
I've been looking into using this on Scala.js as well and I got decoding working locally using this branch like @MirkoManojlovic already said.

Due to missing `java.crypto` and `javax.security` APIs verification does not work on Scala.js (I guess neither on native https://github.com/jwt-scala/jwt-scala/pull/481), but decoding in core and circe is working fine with verification disabled:

```scala
Jwt.decode(
  "<base64encoded jwt>",
  options = JwtOptions(signature = false)
)
```

```scala
JwtCirce.decodeJson(
  "<base64encoded jwt>",
  options = JwtOptions(signature = false, expiration = false)
)
```

If we enable tests on JS, we are running into linking errors as seen here https://github.com/jwt-scala/jwt-scala/pull/453, but some suites like `JwtBase64Spec` and `JwtClaimSpec` are working on JS as well. 

So my suggestion would be to merge this (minus the latest commit).

Then in another PR we could gradually refactor so we can enable all possible tests on JS as well and document the JS restrictions regarding verification. I'd also be happy to help out here.

**Comment by erwan** [comment]
hey, I don't know much about ScalaJS so I'll trust people who know about it - if someone points me to a PR that doesn't break anything outside of ScalaJS and does what you want, I'd be happy to merge it.

**Comment by sbrunk** [comment]
Ok I cherry-picked the first commit from this PR onto current master to trigger the CI: https://github.com/jwt-scala/jwt-scala/pull/482

@massimosiani do you remember why you did [remove JS from the build](https://github.com/jwt-scala/jwt-scala/pull/460/commits/e89929c5a2c2cc9c4884573468e9374458183462) again in the second commit?

**Comment by massimosiani** [comment]
@MirkoManojlovic @sbrunk thanks for trying this. Unfortunately on my project I had to implement a custom check, but willing to introduce this awesome lib back.

The missing packages from the Scala.Js implementation will prevent projects to work properly, without compile-time errors.

One possible solution is to move the verify methods to a Platform trait for the JVM, and the corresponding js and native Platform trait should be empty, so that the JVM modules will compile correctly, while the verify methods won't be available in the js and native artifacts. The same pattern has been extensively employed, for instance, in cats-effect (example [more MapRef methods for jvm](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/std/jvm/src/main/scala/cats/effect/std/MapRefCompanionPlatform.scala) versus [no extension for js and native](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/std/js-native/src/main/scala/cats/effect/std/MapRefCompanionPlatform.scala)).

The same should be done for tests of course, e.g. [shared tests for MapRef](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/tests/shared/src/test/scala/cats/effect/std/MapRefSpec.scala) and [specific tests for the jvm](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/tests/jvm/src/test/scala/cats/effect/std/MapRefJVMSpec.scala).

We could unify the js and native implementation, since neither has verify methods.
Implementing the packages in the Scala.js implementation will take much longer, I think.

What do you think?

**Comment by sbrunk** [comment]
> The missing packages from the Scala.Js implementation will prevent projects to work properly, without compile-time errors.

Yes but if you don't need verification on the client side, it is already usable to extract user/group information etc. from the token. That already works now (see my example above).
 
> One possible solution is to move the verify methods to a Platform trait for the JVM, and the corresponding js and native Platform trait should be empty, so that the JVM modules will compile correctly, while the verify methods won't be available in the js and native artifacts. The same pattern has been extensively employed, for instance, in cats-effect (example [more MapRef methods for jvm](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/std/jvm/src/main/scala/cats/effect/std/MapRefCompanionPlatform.scala) versus [no extension for js and native](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/std/js-native/src/main/scala/cats/effect/std/MapRefCompanionPlatform.scala)).

Agreed that should be helpful to make it more safe and to avoid any linking or even runtime errors when people try to use verification on JS.

> The same should be done for tests of course, e.g. [shared tests for MapRef](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/tests/shared/src/test/scala/cats/effect/std/MapRefSpec.scala) and [specific tests for the jvm](https://github.com/typelevel/cats-effect/blob/e9aeb8ccb53016a37588256f4c8086f8ef1612bc/tests/jvm/src/test/scala/cats/effect/std/MapRefJVMSpec.scala).

I've already started that for the simple cases in #482 (which builds on your work here), moving a few tests that don't require verification into `shared`, and the others into `jvm`.

> We could unify the js and native implementation, since neither has verify methods. Implementing the packages in the Scala.js implementation will take much longer, I think.

Makes sense, adding @sideeffffect as he's started the native port in #481

> What do you think?

My suggestion is to get this in (see #482 for an up-to-date version of this PR) to be able to use decoding in the browser, and then in future PRs try refactoring like you've suggested to make sure you can't accidentally call verification from JS/native until we have a solution for it.

**Comment by erwan** [comment]
merged with #482

**Pull request closed without merging**